aggregator/src/aggregator.rs

```diff
@@ -1115,35 +1126,91 @@ impl VdafOps {
             ));
         }
 
-        // Check that we can decrypt the report. This isn't required by the spec
-        // but this exercises HPKE decryption and saves us the trouble of
-        // storing reports we can't use. We don't inform the client if this
-        // fails.
-        if let Err(error) = hpke::open(
+        // Decode (and in the case of the leader input share, decrypt) the remaining fields of the
+        // report before storing them in the datastore. The spec does not require the /upload
+        // handler to do this, but it exercises HPKE decryption, saves us the trouble of storing
+        // reports we can't use, and lets the aggregation job handler assume the values it reads
+        // from the datastore are valid. We don't inform the client if this fails.
+        let public_share =
+            match A::PublicShare::get_decoded_with_param(&vdaf, report.public_share()) {
+                Ok(public_share) => public_share,
+                Err(err) => {
+                    warn!(
+                        report.task_id = %report.task_id(),
+                        report.metadata = ?report.metadata(),
+                        ?err,
+                        "public share decoding failed",
+                    );
+                    upload_decrypt_failure_counter.add(&Context::current(), 1, &[]);
```

Comment:
I think we should create a separate counter metric for decoding errors, as it will be very useful to track decryption failures separate from other failure cases.